### PR TITLE
get_status() RFC tests

### PR DESCRIPTION
**Pull request by Sami1309**
# Description
Tests described in the RFC for get client status
https://docs.google.com/document/d/1sgaMoRt0fkgPMmHlZzjHYBm9wl2N6Vl5Iilio6fsN_A/edit#

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have fixed any merge conflicts

### CODE REVIEW

File: client/tests/get_status_test.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
+import pytest
+import featureform as ff
+from featureform.resources import TrainingSet, Feature, ResourceColumnMapping
+from featureform.serving import ServingClient
+
+from datetime import timedelta
+
+from .proto import metadata_pb2_grpc, metadata_pb2
+
+def test_pending_training_set_hosted(mocker):
```
Comment (sdreyer) [inline client/tests/get_status_test.py:10]: make your life easier and use @pytest.mark.parametrize

File: client/tests/wait_resource_test.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+def test_wait_trainingset(wait_function):
+    feature = Feature(name="test_name",
+                variant="test_variant")
+    feature.wait_function = wait_function
+    try:
+        feature.wait()
+
+
+def test_waits():
```
Comment (sdreyer) [inline client/tests/wait_resource_test.py:46]: this file will run all the tests then rerun them again in this function

File: client/tests/get_status_test.py (hunk 1 above)
Comment (Sami1309) [reply client/tests/get_status_test.py]: Made it paramaterized

File: client/tests/wait_resource_test.py (hunk 2 above)
Comment (Sami1309) [reply client/tests/wait_resource_test.py]: fixed

File: client/tests/get_status_test.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    ResourceStatus.FAILED: metadata_pb2.ResourceStatus(status=metadata_pb2.FAILED)
+}
+
+@pytest.mark.parametrize("mocker,functions"
+    [(mocker, value) for value in proto_get_functions.values()])
+@pytest.mark.parametrize("status", [(key, status_map[key]) for key in status_map.key()])
+def test_expected_status(mocker, functions, status):
+    mocker.patch(
+        functions["proto_stub"],
+        return_value=[functions["mock_return"](functions["proto_resource"],status[1])]
```
Comment (sdreyer) [inline client/tests/get_status_test.py:54]: Not sure if its just me, but the way this is broken up is really difficult for me to understand whats going on

File: client/tests/get_status_test.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    )
+    assert status[0] == functions["client_function"].get_status()
```
Comment (sdreyer) [inline client/tests/get_status_test.py:56]: What is status[0]?

File: client/tests/get_status_test.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    "label": {
+        "proto_stub": 'metadata_pb2_grpc.ApiStub.GetLabelVariants',
+        "proto_resource", metadata_pb2.Label(name="name", variant="variant", description="description")
+        "mock_return": resource_with_status,
+        "client_function": ResourceClient(host="mock_host", local=False, insecure=True, cert_path="").get_label("",""),
+    }
+}
+
+status_map = {
+    ResourceStatus.CREATED: metadata_pb2.ResourceStatus(status=metadata_pb2.CREATED),
```
Comment (sdreyer) [inline client/tests/get_status_test.py:42]: Where did this dict come from?

File: client/tests/get_test.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  rc_get_functions_var = {
+    "source_var": [rc.get_source, "transactions", "kaggle"],
+    "feature_var": [rc.get_feature, "avg_transactions", "quickstart"],
+    "label_var": [rc.get_label, "fraudulent", "quickstart"],
+    "training_set_var": [rc.get_training_set, "fraud_training", "quickstart"]
+  }
+
+  with patch('sys.stdout', new = StringIO()) as fake_out:
+    if resource_type in rc_get_functions:
+      value = rc_get_functions[resource_type][0](rc_get_functions[resource_type][1])
```
Comment (sdreyer) [inline client/tests/get_test.py:33]: This is super confusing

File: client/tests/get_status_test.py (hunk 5 above)
Comment (Sami1309) [reply client/tests/get_status_test.py]: removed

File: client/tests/get_test.py (hunk 6 above)
Comment (Sami1309) [reply client/tests/get_test.py]: Changed code to make less confusing

File: client/src/featureform/cli.py [hunk 7]
```diff
@@ -187,4 +186,4 @@ def apply(host, cert, insecure, local, files, dry_run):
 
 
 if __name__ == '__main__':
-    cli()
+    cli()
```
Comment (sdreyer) [inline client/src/featureform/cli.py:189]: Whats going on here?

File: client/tests/get_status_test.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+from .proto import metadata_pb2_grpc, metadata_pb2
+
+def resource_with_status(resource, status):
+    resource.status = status
+    return resource
+
+expected_list = [("PENDING", ff.ResourceStatus.PENDING), ("READY", ff.ResourceStatus.READY), ("CREATED", ff.ResourceStatus.CREATED), ("FAILED", ff.ResourceStatus.FAILED)]
+
+@pytest.mark.paramaterize(input, expected, expected_list)
```
Comment (sdreyer) [inline client/tests/get_status_test.py]: Should check the formatting for this. I think its like @pytest.mark.paramaterize("input,expected", expected_list)

File: client/tests/wait_resource_test.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+wait_functions = {
+    "success": wait_function_success,
+    "failure": wait_function_failure,
+}
+
+@pytest.mark.parametrize("resource", [resource for resource in test_resources.values()])
+def test_wait_success(resource):
+    resource.wait_function = wait_function_success
+    try:
```
Comment (sdreyer) [inline client/tests/wait_resource_test.py:65]: Doesnt this need an except?

File: client/tests/get_test.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
+
```
Comment (sdreyer) [inline client/tests/get_test.py:1]: Delete the old file

File: client/tests/print_resource_test.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+import featureform as ff
+from featureform.resources import TrainingSet, Feature, ResourceColumnMapping
+from featureform.serving import ServingClient
+
+from .test_outputs import print_outputs
+
+from datetime import timedelta
+
+def test_feature_print():
+    feature = print_outputs["feature"]["unit_test"]["resource"]
```
Comment (sdreyer) [inline client/tests/print_resource_test.py:11]: rename "resource" to "object" or "resource_object". Its unclear what this is fetching

Comment (sdreyer) [reply client/tests/print_resource_test.py:11]: You may want to use fixtures to make it more clear
```
@pytest.fixture
def feature_object():
      return print_outputs["feature"]["unit_test"]["object"]

@pytest.fixture
def feature_expected_print():
      return print_outputs["feature"]["unit_test"]["output"]

def test_feature_print(feature_object, feature_expected_print):
       assert feature_expected_print == f"{feature_object}"
```

File: client/tests/wait_resource_test.py (hunk 9 above)
Comment (Sami1309) [reply client/tests/wait_resource_test.py]: It's meant to run with no error. Should I just run it without an except, or check the return (it just returns itself)

File: client/tests/print_resource_test.py (hunk 11 above)
Comment (Sami1309) [reply client/tests/print_resource_test.py]: changed to fixture schema

File: client/src/featureform/cli.py (hunk 7 above)
Comment (Sami1309) [reply client/src/featureform/cli.py:189]: I removed the changes to this file so it's the same as in the branch I'm merging into, this is just a weird spurious diff.

File: client/src/featureform/cli.py [hunk 12]
```diff
@@ -35,7 +35,6 @@ def cli():
     |  _/ _ \/ _` | __| | | | '__/ _ \  _/ _ \| '__| '_ ` _ \ 
     | ||  __/ (_| | |_| |_| | | |  __/ || (_) | |  | | | | | |
     \_| \___|\__,_|\__|\__,_|_|  \___|_| \___/|_|  |_| |_| |_|
-
```
Comment (sdreyer) [inline client/src/featureform/cli.py:38]: Nitpick but this should be left alone

**Review by sdreyer — APPROVED** [review]

**Pull request merged**